Update dictionary-expressions.md (#9297)
dotnet/csharplang: proposals/dictionary-expressions.md

In addition, the value must implement the nongeneric `IDictionary` interface. This enables collection expressions to support dynamic introspection in scenarios such as data binding.

A compliant implementation is free to:

1. Use an existing type that implements the required interfaces.
2. Synthesize a type that implements the required interfaces.

In either case, the type used is allowed to implement a larger set of interfaces than those strictly required.

Synthesized types are free to employ any strategy they want to implement the required interfaces properly. For example, returning a cached singleton for empty collections, or a synthesized type which inlines the keys/values directly within itself, avoiding the need for additional internal collection allocations.

> The value must return true when queried for `ICollection<T>.IsReadOnly`. This ensures consumers can appropriately tell that the collection is non-mutable, despite implementing the mutable views.  The value must throw on any call to a mutation methods (like `IDictionary<TKey, TValue>.Add`). This ensures safety, preventing a non-mutable collection from being accidentally mutated.

Author: ```suggestion
1. The value must return `true` when queried for `ICollection<T>.IsReadOnly`. This ensures consumers can appropriately tell that the collection is non-mutable, despite implementing the mutable views.
1. The value must throw on any call to a mutation methods (like `IDictionary<TKey, TValue>.Add`). This ensures safety, preventing a non-mutable collection from being accidentally mutated.
```

## Interface translation

### Non-mutable interface translation

Given a target type `IReadOnlyDictionary<TKey, TValue>`, a compliant implementation is required to produce a value that implements that interface. If a type is synthesized, it is recommended the synthesized type implements `IDictionary<TKey, TValue>` as well. This ensures maximal compatibility with existing libraries, including those that introspect the interfaces implemented by a value in order to light up performance optimizations.

> In addition, the value must implement the nongeneric `IDictionary` interface. This enables collection expressions to support dynamic introspection in scenarios such as data binding.

Reviewer: Is there a specific reason we are requiring the runtime type implements `IDictionary` but not `IDictionary<TKey, TValue>`? Maybe because we have a stronger motivating scenario for the former than the latter?

Author: for now, i'm following the same rules/verbiage as what we did for lists and CEs.    

That said, i'm discussing with Dustin, and some of our winforms/designer people if this same motivation exists here.  I will add this as an open question.  But it comes down to.

1. should we just do what we did for CEs, with a direct analog for each part?  (this is the spec as above)
2. should we carve out some of these pieces if there is legit no ask?

'1' seems safer, at the cost of a bit more work.  But we have experience with that from when we did CEs.  So it's path of least resistance.

'2' lowers costs.  But does potentially make us sad later if we ship without this and then someone uses this and runs to the analogous case.

I also prefer '1' because it becomes less around special rules/exceptions, and more about "dictionaries just work like normal CEs".  As much as they can align, i prefer it, because now there's less potential 'gotchas'.  Basically, we get to say: 

> It's virtually the same. Except that if your element type is `KVP<,>` and you have an indexer, we'll think of you as a dict, and better handle arbitrary KVPs and will use the indexer to initialize.

I *really* like that, and would prefer alignment *unless* motivated to not have it.

Author: added a new open question at the bottom for this.

Reviewer: The fact that this is in alignment with list is enough for me. Basically if `IReadOnlyList<T>` implements `IList` but not `IList<T>` then yes I would want this to work the same. We shipped that, so, if someone is not happy with it, then presumably they can come to us, and we could discuss changing both lists and dictionaries at the same time.

Author: Right.  And, frankly, it feels liek the sequential collection types would be much more of a big deal, if there was an issue.  Give nothing seemingly wrong there, aligning with it feels good to me.

A compliant implementation is free to:

> 1. Use an existing type that implements the required interfaces.
> 2. Synthesize a type that implements the required interfaces.

Reviewer 2: ```suggestion
1. Use an existing BCL type that implements the required interfaces.
2. Synthesize a type that implements the required interfaces.
```

In either case, the type used is allowed to implement a larger set of interfaces than those strictly required.

Synthesized types are free to employ any strategy they want to implement the required interfaces properly. For example, returning a cached singleton for empty collections, or a synthesized type which inlines the keys/values directly within itself, avoiding the need for additional internal collection allocations.

> 1. The value must return `true` when queried for `ICollection<T>.IsReadOnly`. This ensures consumers can appropriately tell that the collection is non-mutable, despite implementing the mutable views.
> 1. The value must throw on any call to a mutation methods (like `IDictionary<TKey, TValue>.Add`). This ensures safety, preventing a non-mutable collection from being accidentally mutated.

Reviewer 2: ```suggestion
1. The value must return `true` when queried for `ICollection<T>.IsReadOnly`. This ensures consumers can appropriately tell that the collection is non-mutable, despite implementing the mutable views.
1. The value must throw on any call to a mutation methods (like `IDictionary<TKey, TValue>.Add`). This ensures safety, preventing a non-mutable collection from being accidentally mutated.
```

> 1. The value must return `true` when queried for `ICollection<T>.IsReadOnly`. This ensures consumers can appropriately tell that the collection is non-mutable, despite implementing the mutable views.
> 1. The value must throw on any call to a mutation methods (like `IDictionary<TKey, TValue>.Add`). This ensures safety, preventing a non-mutable collection from being accidentally mutated.

Reviewer 2: ```suggestion
1. The value must return `true` when queried for `ICollection<T>.IsReadOnly`. This ensures consumers can appropriately tell that the collection is non-mutable, despite implementing the mutable views.
1. The value must throw on any call to a mutation method (like `IDictionary<TKey, TValue>.Add`). This ensures safety, preventing a non-mutable collection from being accidentally mutated.
```

> 1. The value must return `true` when queried for `ICollection<T>.IsReadOnly`. This ensures consumers can appropriately tell that the collection is non-mutable, despite implementing the mutable views.
> 1. The value must throw on any call to a mutation methods (like `IDictionary<TKey, TValue>.Add`). This ensures safety, preventing a non-mutable collection from being accidentally mutated.

Reviewer 2: ```suggestion
Regardless of whether the implementation type is a well-known type or a synthesized type:
1. The implementation must return `true` when queried for `ICollection<T>.IsReadOnly`. This ensures consumers can appropriately tell that the collection is non-mutable, despite implementing the mutable views.
1. The implementation must throw on any call to a mutation method (like `IDictionary<TKey, TValue>.Add`). This ensures safety, preventing a non-mutable collection from being accidentally mutated.
```

> 1. The value must throw on any call to a mutation methods (like `IDictionary<TKey, TValue>.Add`). This ensures safety, preventing a non-mutable collection from being accidentally mutated.

Reviewer 2: ```suggestion
1. The value must throw on any call to a mutation method (like `IDictionary<TKey, TValue>.Add`). This ensures safety, preventing a non-mutable collection from being accidentally mutated.
```

**Resolution:** Use *indexer* as the lowering form. [LDM-2024-03-11](https://github.com/dotnet/csharplang/blob/main/meetings/2024/LDM-2024-03-11.md#conclusions)

### Answered question 5

What types and translation should be used when targeting dictionary interfaces (`IDictionary<TKey, TValue>` or `IReadOnlyDictionary<TKey, TValue>`)?

> **Resolution:** Use the same rules used for mutable and read-only interfaces for normal

Reviewer 2: >read-only [](http://example.com/codeflow?start=56&length=9)

non-mutable?

> **Resolution:** Use the same rules used for mutable and read-only interfaces for normal

Reviewer 2: ```suggestion
**Resolution:** Use the same rules used for mutable and non-mutable interfaces for normal
```

> **Resulution:** Use the same rules used for mutable and read-only interfaces for normal
> [*collection expressions*](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-12.0/collection-expressions.md#interface-translation)
> analogously translated to dictionaries.  Full details can be found in [interface-translation](Interface Translation).  

Reviewer 2: ```suggestion
**Resolution:** Use the same rules used for mutable and read-only interfaces for normal
[*collection expressions*](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-12.0/collection-expressions.md#interface-translation)
analogously translated to dictionaries.  Full details can be found in [interface-translation](#interface-translation).  
```

> [LDM-2025-04-09](link-tbd)

Reviewer 2: ```suggestion
[LDM-2025-04-09](https://github.com/dotnet/csharplang/blob/main/meetings/2025/LDM-2025-04-09.md#conclusion)
```

## Open Questions

### Question: Parsing ambiguity

Parsing ambiguity around: `[a ? [b] : c]`

Working group recommendation: Use normal parsing here.  So this would be the same as `[a ? ([b]) : (c)]` (a collection expression containing a conditional expression).
If the user wants a `key_value_pair_element` here, they can write: `[(a?[b]) : c]`.  This code already will exist today in a collection expression, and it should not
change meaning.

### Question: Implement non-generic `IDictionary` when targeting `IReadOnlyDictionary<,>`

Collection expressions specified explicitly:

> Given a target type which does not contain mutating members, namely `IEnumerable<T>`, `IReadOnlyCollection<T>`, and `IReadOnlyList<T>`, a compliant implementation is required to produce a value that implements that interface. ...
>
> In addition, the value must implement the nongeneric `ICollection` and `IList` interfaces. This enables collection expressions to support dynamic introspection in scenarios such as data binding.

Do we want a similar correspondance when the target type is `IReadOnlyDictionary<,>`?  Specifically, should the value be required to implement the non-generic `IDictionary` interface?

> Working group recomendation: Yes.  Implement `IDictionary`.  Specifically, all existing non-mutable dictionary types the compiler might use (`ReadOnlyDictionary<,>`, `FrozenDictionary<,>` or `ImmutableDictionary<,>`) all implement `IDictionary` (despite not needing to).  It is normal for our non-mutable types to expose this interface, and requiring it ensures maximal compatibility with any existing consumption code, with minor cost to the compiler.  For the most part, in practice the compiler will just be using these types anyways (most likely `ReadOnlyDictionary<,>` and `FrozenDictionary<,>`), so the guarantee is no actual cost.

Reviewer 2: ```suggestion
Working group recomendation: Yes.  Implement `IDictionary`.  Specifically, all existing non-mutable dictionary types the compiler might use (`ReadOnlyDictionary<,>`, `FrozenDictionary<,>` or `ImmutableDictionary<,>`) all implement `IDictionary` (despite not needing to).  It is normal for our non-mutable types to expose this interface, and requiring it ensures maximal compatibility with any existing consumption code.
```
